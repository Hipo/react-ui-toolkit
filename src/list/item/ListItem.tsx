import "./_list-item.scss";

import React from "react";
import classNames from "classnames";

import {KEYBOARD_EVENT_KEY} from "../../core/utils/keyboard/keyboardEventConstants";

interface ListItemProps {
  children?: React.ReactNode;
  testid?: string;
  customClassName?: string;
  clickableListItemProps?: {
    onClick: React.ReactEventHandler;
    tabIndex?: number;
  };
  id?: string;
  role?: string;
  ariaSelected?: boolean;
  listItemRef?: React.RefObject<HTMLLIElement>;
}

function ListItem({
  id,
  testid,
  children,
  customClassName,
  clickableListItemProps,
  role,
  ariaSelected,
  listItemRef
}: ListItemProps) {
  const containerClassName = classNames("list-item", customClassName);
  let listItem = (
    <li
      ref={listItemRef}
      id={id}
      data-testid={testid}
      className={containerClassName}
      role={role}
      aria-selected={ariaSelected}>
      {children}
    </li>
  );

  if (clickableListItemProps) {
    listItem = (
      <li ref={listItemRef} id={id} data-testid={testid} className={containerClassName}>
        <div
          role={"button"}
          tabIndex={clickableListItemProps.tabIndex || 0}
          className={"list-item__click-wrapper"}
          onClick={handleClick}
          onKeyPress={handleKeyPress}>
          {children}
        </div>
      </li>
    );
  }

  return listItem;

  function handleClick(event: React.SyntheticEvent) {
    clickableListItemProps!.onClick(event);
  }

  function handleKeyPress(event: React.KeyboardEvent) {
    if (event.key === KEYBOARD_EVENT_KEY.ENTER) {
      handleClick(event);
    }
  }
}

export default ListItem;
