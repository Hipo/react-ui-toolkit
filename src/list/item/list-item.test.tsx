import React from "react";
import {render, cleanup, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import {testA11y} from "../../core/utils/test/testUtils";
import ListItem, {ListItemProps} from "./ListItem";

describe("<ListItem />", () => {
  afterEach(cleanup);

  const defaultListItemProps: ListItemProps = {
    testid: "list-item",
    children: <p data-testid={"list-item.content"}>{"Test"}</p>
  };

  it("should render correctly", () => {
    render(<ListItem {...defaultListItemProps} />);
  });

  it("should matches snapshot", () => {
    const tree = create(<ListItem {...defaultListItemProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should pass a11y test", async () => {
    const {container} = render(<ListItem {...defaultListItemProps} />);

    await testA11y(container, {rules: {listitem: {enabled: false}}});
  });

  it("should render children correctly", () => {
    const {getByTestId} = render(<ListItem {...defaultListItemProps} />);

    expect(getByTestId("list-item")).toContainElement(getByTestId("list-item.content"));
  });

  it("clickableListItemProps should work correctly", () => {
    const handleClick = jest.fn();
    const {getByTestId, getByRole} = render(
      <ListItem
        clickableListItemProps={{onClick: handleClick, tabIndex: 2}}
        {...defaultListItemProps}
      />
    );
    const listItemButton = getByRole("button");

    fireEvent.click(listItemButton);

    expect(getByTestId("list-item")).toContainElement(listItemButton);
    expect(listItemButton).toHaveAttribute("tabIndex", "2");
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("should run click event handler when enter key pressed", () => {
    const handleClick = jest.fn();
    const {getByRole} = render(
      <ListItem
        clickableListItemProps={{onClick: handleClick, tabIndex: 2}}
        {...defaultListItemProps}
      />
    );

    fireEvent.keyPress(getByRole("button"), {
      key: "Enter",
      code: "Enter",
      keyCode: 13
    });

    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
