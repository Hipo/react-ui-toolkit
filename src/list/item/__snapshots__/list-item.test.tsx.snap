// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> should matches snapshot 1`] = `
<li
  className="list-item"
  data-testid="list-item"
>
  <p
    data-testid="list-item.content"
  >
    Test
  </p>
</li>
`;
